api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+
+	// AnnotationFilter describes an annotation filter
+	// used to filter which source instance resources
+	// ExternalDNS publishes records for.
+	// The annotation filter uses label selector semantics
+	// against source resource annotations.
+	//
+	// +optional
+	AnnotationFilter map[string]string `json:"annotationFilter,omitempty"`
+
+	// Namespace instructs ExternalDNS to only acknowledge
+	// source resource instances in a specific namespace.
+	//
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// Service describes source configuration options specific
+	// to the service source resource.
+	//
+	// +optional
+	Service *ExternalDNSServiceSourceOptions `json:"service,omitempty"`
+
+	// CRD describes source configuration options specific
+	// to the CRD source resource.
+	//
+	// +optional
+	CRD *ExternalDNSCRDSourceOptions `json:"crd,omitempty"`
+
+	// OpenShiftRoute source currently has no unique configuration options.
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=Ignore;Allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the ExternalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// PublishPolicy determines what kinds of Service
+	// resources are published to the given DNS provider.
+	//
+	// The following publishing policies are available:
+	//
+	//  "PublishExternal": Only publish DNS records for
+	//  service that are externally reachable,
+	//  such as NodePort, ExternalName, and LoadBalancer
+	//  services. This is the default behavior of ExternalDNS.
+	//
+	//  "PublishInternalExternal": Publish DNS records
+	//   for externally reachable services, in addition to
+	//   internally reachable services, such as ClusterIP
+	//   services.
+	//
+	// +optional
+	PublishPolicy *ServicePublishPolicy `json:"publishPolicy,omitempty"`
+
+	// ServiceType determines what types of Service resources
+	// are watched by ExternalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP" (requires a publishPolicy of "PublishInternalExternal")
+	//
+	// One or more Service types can be specified, if desired.
+	//
+	// If no service types are provided, ExternalDNS will be
+	// configured to create DNS records for LoadBalancer services
+	// only by default.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:={"LoadBalancer"}
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	ServiceType []corev1.ServiceType `json:"serviceType,omitempty"`
```

Comment:
Why have this option in addition to publishpolicy?  This appears to cover all the cases that PublicExternal does with better granularity.

Reply:
In retrospect, im not sure why I thought adding an unnecessary publish policy was a good idea.

I've removed the publish policy field and elaborated on the implications of using the ClusterIP service type in the comments above.